Fix/extract csrf token from request

meschac38700 opened the pull request:
This pull request addresses a regression introduced by the recent flexible CSRF validation feature. 
It fixes the correct extraction of CSRF tokens from both JSON and form-data requests.

### Checklist
  - [x] CSRF tokens can be extracted from headers, JSON, and form-data requests
  - [x] All relevant tests pass

meschac38700 committed:
feat: add python-multipart as new dependency

meschac38700 committed:
tests: cover extract csrf token from requests

meschac38700 committed:
fix: csrf token extraction for form-data and JSON requests

meschac38700 committed:
feat: remove python-multipart dependency

meschac38700 committed:
fix: first attempt using json.loads

aekasitt commented:
No more changes right?

aekasitt committed:
refactor(fmt): format by ruff; standardize imports

aekasitt committed:
test(flags): define --flexible & --normal flags

aekasitt committed:
test: update reasons for manual skipping

meschac38700 commented:
@aekasitt 
I understand, but the behavior when extracting the token from the request data differs between the flexible and normal versions. This PR does not exactly reimplement exactly the same code as in the base package, but adds support for extracting the token from the form/data and JSON data of a request.

#### in normal package core.py(validate_csrf(...))
```Python
    if self._token_location == "header":
      token = self.get_csrf_from_headers(request.headers)
    else:
      if hasattr(request, "_json") and request._json is not None:
        token = request._json.get(self._token_key, "")
      elif hasattr(request, "_form") and request._form is not None:
        form_data: Union[None, UploadFile, str] = request._form.get(self._token_key)
        if not form_data or isinstance(form_data, UploadFile):
          raise MissingTokenError("Form data must be of type string")
        token = form_data
      else:
        token = self.get_csrf_from_body(await request.body())
```

aekasitt commented:
Check the latest changes on `master`

meschac38700 commented:
> Check the latest changes on `master`

@aekasitt 
I understand. However, i've noticed that in the flexible mode, we lose the ability to process data sent in the request body with "application/json" content type. 
This is crucial for librairies like *Pydantic*.
For instance, consider the following scenario:


```Python
# tests.flexible.__init__.py

...

@fixture
def flexible_client() -> Generator[TestClient, None, None]:
    ...
    app: FastAPI = FastAPI()
    from pydantic import BaseModel

    class User(BaseModel):
        name: str
    
    @app.post("/protected", response_class=JSONResponse)
    async def update_resource(
         request: Request,
         csrf_protect: Annotated[CsrfProtect, Depends(CsrfProtect)],
         user: User
    ) -> JSONResponse:
        await csrf_protect.validate_csrf(request)      
        response: JSONResponse = JSONResponse(status_code=200, content={"detail": "OK", **user.model_dump()})
        csrf_protect.unset_csrf_cookie(response)  # prevent token reuse
        return response    
```    
This endpoint specifically expects data in a JSON format to properly validate the User object.
Unfortunately, using `data=...` in our tests sends `application/x-www-form-urlencoded (or raw body if bytes/str)` This is what's expected by endpoints using `Form(...)`.
As a result, Pydantic fails with the following error: 
```Console
(Pdb) response.json ()
{'detail': [{'type': 'model_attributes_type', 'loc': ['body'], 'msg': 'Input should be a valid dictionary or object to extract fields from', 'input': 'csrf-token=ea4a6556c0a73b4564c4f06ee9828b3b41081c12&name=Foo'}]}
(Pdb) 
```
When i use `json=...` instead of `data=...` i get a different error. This one is related to the processing within the `flexible.core` module because it doesn't know how to parse the data in JSON format.
<img width="1525" height="797" alt="image" src="https://github.com/user-attachments/assets/91fa0b43-00ed-4869-a3a4-0368a6f7b998" />

aekasitt commented:
I changed the encoding for test form submissions in these test modules
- `tests/cookie_form.py`
- `tests/flexible/cookie_form.py`
as such:

```diff
+ from urllib.parse import urlencode
...
+   headers: dict[str, str] = {"Content-Type": "application/x-www-form-urlencoded"}
+   payload: dict[str, str] = {"csrf-token": csrf_token} if csrf_token is not None else {}
+   content: bytes = urlencode(payload).encode("utf-8")
+   response = test_client.post("/protected", content=content, headers=headers)
```
Does this work on your end?

meschac38700 commented:
> I changed the encoding for test form submissions in these test modules
> 
>     * `tests/cookie_form.py`
> 
>     * `tests/flexible/cookie_form.py`
>       as such:
> 
> 
> ```diff
> + from urllib.parse import urlencode
> ...
> +   headers: dict[str, str] = {"Content-Type": "application/x-www-form-urlencoded"}
> +   payload: dict[str, str] = {"csrf-token": csrf_token} if csrf_token is not None else {}
> +   content: bytes = urlencode(payload).encode("utf-8")
> +   response = test_client.post("/protected", content=content, headers=headers)
> ```
> 
> Does this work on your end?

Unfortunately it doesn't work either, I get the following Pydantic error: 

```Console
{'detail': [{'type': 'json_invalid', 'loc': ['body', 0], 'msg': 'JSON decode error', 'input': {}, 'ctx': {'error': 'Expecting value'}}]}
```

And with the **{"Content-Type": "application/x-www-form-urlencoded"}** header, it's not better

```Console
{'detail': [{'type': 'model_attributes_type', 'loc': ['body'], 'msg': 'Input should be a valid dictionary or object to extract fields from', 'input': 'csrf-token=0cd14e39f99ffe0ea240b20e6efc9d432a2da9f9&name=Foo'}]}
```

### Here's my test:
I updated "test_submit_csrf_token_in_body_and_cookie" (tests/flexible/cookie_body.py)
<img width="850" height="299" alt="image" src="https://github.com/user-attachments/assets/038d71fa-f57a-4985-b206-3067626f522b" />


And the flexible_client. (tests/flexible/__init__.py)

<img width="871" height="641" alt="image" src="https://github.com/user-attachments/assets/343e658f-2495-4bc3-a4a8-bbd7b1d8daba" />

You can test it on your end. @aekasitt

aekasitt commented:
@meschac38700 

I understand your concern, but parsing Form into Pydantic model is beyond the scope of this library but falls upon the FastAPI implementation direction. Please join the discussions at [Add Support for Pydantic models in multipart/form data requests.](https://github.com/fastapi/fastapi/discussions/8406)

---
PS. If submission of FormData into Pydantic's custom model prematurely consumes attached csrf-token from this library, we should both jump to task and make it work.

meschac38700 commented:
> @meschac38700
> 
> I understand your concern, but parsing Form into Pydantic model is beyond the scope of this library but falls upon the FastAPI implementation direction. Please join the discussions at [Add Support for Pydantic models in multipart/form data requests.](https://github.com/fastapi/fastapi/discussions/8406)
> 
> PS. If submission of FormData into Pydantic's custom model prematurely consumes attached csrf-token from this library, we should both jump to task and make it work.

Hi @aekasitt, thanks for your feedback.
i think there might be a misunderstanding here.
My fix is not about submitting FormData into a Pydantic model.

The issue only concerns the flexible submodule:
    - In the main library, CSRF token extraction works correctly from both `application/json` and `multipart/form-data`, as well through manual parsing `get_csrf_from_body(..)`
    - In the flexible submodule, however, only form-data and manual parsing are supported `application/json` requests are not handled, which breaks API-first workflows

My PR specifically fixes this inconsistency so that flexible behaves the same as the rest of the libray, by allowing token extraction from `application/json` bodies as expected.

Could you please re-check under this angle? I believe this change aligns with the intended behavior of the library rather than extending it beyond scope.

### To better illustrate the issue:

```Python
from typing import Annotated

from fastapi import Depends, FastAPI
from pydantic import BaseModel
from fastapi.responses import JSONResponse

from fastapi_csrf_protect.flexible import CsrfProtect


class User(BaseModel):
    name: str


app = FastAPI()

@app.post("/users/", response_class=JSONResponse)
def create_user(csrf_protect: Annotated[CsrfProtect, Depends(CsrfProtect)], user: User):
    ...
```

Now, two scenarios happen:

## 1. AJAX request with application/json body

```Python
{"csrf_token": "****", "name": "Foo"}
```

    - Pydantic correctly parses the payload and initializes user.
    - But the flexible submodule crashes because it doesn't know how to extract `csrf_token` from a JSON body (it only expects x-www-form-urlencoded).

## 2 . Form submission with application/x-www-form-urlencoded body**

```Python
csrf_token=****&name=Foo
```

    - CSRF extraction works in this case
    - But then Pydantic fails, since it expects JSON for the user model

It other words:

    - The main library already supports both JSON and form-data consistently
    - The flexible submodule is the inconsistent part, because it only handles form-data (and manual parsing), but not JSON.

aekasitt commented:
I readded the pre-verification of content before get_csrf_token_from_body, does this match your desired output?

meschac38700 commented:
> I readded the pre-verification of content before get_csrf_token_from_body, does this match your desired output?

Yes, thank you, that should be fine. At least there is consistency between the two modules.
